walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,93 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 
 public final class FindMeetingQuery {
+
+  /**
+   * This function returns a list of potential time ranges for a meeting 
+   * during the day. The request contains the attendees and the duration
+   * of the meeting, and events contains the other events that are 
+   * scheduled already. If an attendee of the meeting request is an
+   * attendee of another event, the times that the other event covers
+   * are no longer possible for the meeting.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    
+    // If the request is over a day, then there are no time slots available
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return possibleTimes;
+    }
+
+    // Start with the whole day available
+    possibleTimes.add(TimeRange.WHOLE_DAY);
+
+    if (request.getAttendees().size() == 0) {
+        return possibleTimes;
+    }
+
+    // Only update times for events that have attendees from the meeting request
+    for (Event event : events) {
+        for (String attendee : request.getAttendees()) {
+            if (event.getAttendees().contains(attendee)) {
+                updatePossibleTimes(possibleTimes, event.getWhen(), request.getDuration());
+            }
+        }
+    }
+
+    return possibleTimes;
+  }
```

Comment:
What is the time-complexity of this approach? Assume that there are N events and M attendees.

Reply:
I guess this would be O(NM) because it is searching for M attendees for each event, and the lookup is constant time for a HashSet of attendees. I wasn't sure if there was a better way of doing this

Reply:
I think it takes O(N) time to both add and remove an element from an array list, in which case your algorithm might actually be O(N^4). I think its fine to leave it like this for now unless the task is to have an optimized algorithm.

Reply:
would you suggest using a different data structure in order to reduce the complexity?

Reply:
Using a Set would definitely reduce the time to add and remove from O(N) to O(log N), so you can consider doing that.

Reply:
Adding to what Sidharth said -- TreeSets also ensure that your time ranges are sorted, which can help with all the range overlap lookups you're doing.

Right now, you're doing O(N) lookups to find the right place to split. Once this is sorted, though, you can essentially do binary searches to find overlapping time ranges (possible time ranges are always non-overlapping, iiuc, and in a strictly increasing order)

Possibly TreeSet.subSet could get you all the overlapping time ranges (you'd essentially have to do some comparator trickery, and the potential for off by one errors abounds). I assume it's O(logn) and binary searching.

Reply:
^ was very much not meant as a blocking comment, but just something to think about. ;)

Another possibility might be to see if there is preprocessing you can do to make the problem less complex. You aren't required to necessarily work with the data structures and orderedness you're given in the parameters.

For instance, if you had two sorted lists/sets of timeranges, possibleTimes and timeRangesToRemove, would it be faster to remove the time ranges then (in the same way that removeAll() can be faster than individual removes)?